reactor-core/src/main/java/reactor/core/publisher/FluxSwitchOnFirst.java

```diff
@@ -19,610 +19,1029 @@
+			}
+		}
+
+		@Override
+		@SuppressWarnings("unchecked")
+		public boolean tryOnNext(T t) {
+			if (this.done) {
+				Operators.onNextDropped(t, currentContext());
+				return false;
+			}
+
+			final T f = this.firstValue;
+			if (f == null) {
+				this.firstValue = t;
+
+				final long state = markFirstValueReceived(this);
+				if (hasInboundCancelled(state)) {
+					Operators.onDiscard(t, this.outer.currentContext());
+					return true;
+				}
+
+				final Publisher<? extends R> result;
+				final CoreSubscriber<? super R> o = this.outer;
+
+				try {
+					result = Objects.requireNonNull(this.transformer.apply(Signal.next(t,
+							o.currentContext()), this),
+							"The transformer returned a null value");
+				}
+				catch (Throwable e) {
+					this.done = true;
+					o.onError(Operators.onOperatorError(this.s,
+							e,
+							t,
+							o.currentContext()));
+					return false;
+				}
+
+				result.subscribe(o);
+				return true;
+			}
+
+			return ((Fuseable.ConditionalSubscriber<? super T>) this.inner).tryOnNext(t);
+		}
+
+		@Override
+		@SuppressWarnings("unchecked")
+		boolean tryOnNext(CoreSubscriber<? super T> actual, T t) {
+			return ((Fuseable.ConditionalSubscriber<? super T>) actual).tryOnNext(t);
+		}
+	}
+
+	static final class SwitchOnFirstControlSubscriber<T>
+			extends Operators.DeferredSubscription
+			implements InnerOperator<T, T>, ControlSubscriber<T> {
+
+		final AbstractSwitchOnFirstMain<?, T> parent;
+		final CoreSubscriber<? super T>       delegate;
+		final boolean                         cancelSourceOnComplete;
+
+		boolean done;
+
+		SwitchOnFirstControlSubscriber(AbstractSwitchOnFirstMain<?, T> parent,
+				CoreSubscriber<? super T> delegate,
+				boolean cancelSourceOnComplete) {
+			this.parent = parent;
+			this.delegate = delegate;
+			this.cancelSourceOnComplete = cancelSourceOnComplete;
+		}
+
+		@Override
+		public void sendSubscription() {
+			delegate.onSubscribe(this);
+		}
+
+		@Override
+		public void onSubscribe(Subscription s) {
+			if (set(s)) {
+				long previousState = markOutboundSubscribed(this.parent);
+
+				if (hasOutboundCancelled(previousState)) {
+					s.cancel();
+				}
+			}
+		}
+
+		@Override
+		public CoreSubscriber<? super T> actual() {
+			return this.delegate;
+		}
+
+		@Override
+		public void onNext(T t) {
+			if (this.done) {
+				Operators.onNextDropped(t, currentContext());
+				return;
+			}
+
+			this.delegate.onNext(t);
+		}
+
+		@Override
+		public void onError(Throwable throwable) {
+			if (this.done) {
+				Operators.onErrorDropped(throwable, currentContext());
+				return;
+			}
+
+			this.done = true;
+
+			final AbstractSwitchOnFirstMain<?, T> parent = this.parent;
+			long previousState = markOutboundTerminated(parent);
+
+			if (hasOutboundCancelled(previousState) || hasOutboundTerminated(previousState)) {
+				return;
+			}
+
+			if (!hasInboundCancelled(previousState) && !hasInboundTerminated(previousState)) {
```

Comment:
can you remind me what's the operator logic here with `cancelAndError`? I'm not sure I remember what the `andError` implies...

Reply:
This is a corner case scenario related to the #2598. What that one is basically doing if cancellation comes for the outbound, it calls a specific inbound's method which along with upstream cancellation sends a CancellationException to the downstream to ensure that the inboud subscriber is really termianted